File: packages/sdk-typescript/test/client.test.ts
import { describe, it, expect, vi, beforeEach } from 'vitest';
import {
  TinyHumansMemoryClient,
  TinyHumansError,
  InsertMemoryParams,
  QueryMemoryParams,
  DeleteMemoryParams,
  RecallMemoryParams,
  RecallMemoriesParams,
} from '../src/index';

describe('TinyHumansMemoryClient', () => {
  const baseUrl = 'https://api.test.example';
  const token = 'test-token';

  let fetchMock: ReturnType<typeof vi.fn>;

  beforeEach(() => {
    fetchMock = vi.fn();
    vi.stubGlobal('fetch', fetchMock);
  });

  describe('constructor', () => {
    it('throws when token is missing', () => {
      expect(() => new TinyHumansMemoryClient({ token: '' })).toThrow('token is required');
      expect(() => new TinyHumansMemoryClient({ token: '   ' })).toThrow('token is required');
    });

    it('uses provided baseUrl and token', () => {
      const client = new TinyHumansMemoryClient({ token, baseUrl });
      expect(client).toBeDefined();
    });
  });

  describe('insertMemory', () => {
    it('validates required fields', async () => {
      const client = new TinyHumansMemoryClient({ token, baseUrl });
      await expect(client.insertMemory({ title: '', content: 'x', namespace: 'ns', documentId: 'doc_x' })).rejects.toThrow(
        'title is required'
      );
      await expect(client.insertMemory({ title: 't', content: '', namespace: 'ns', documentId: 'doc_x' })).rejects.toThrow(
        'content is required'
      );
      await expect(client.insertMemory({ title: 't', content: 'c', namespace: '', documentId: 'doc_x' })).rejects.toThrow(
        'namespace is required'
      );
      await expect(client.insertMemory({ title: 't', content: 'c', namespace: 'ns', documentId: '' })).rejects.toThrow(
        'documentId is required'
      );
    });

    it('POSTs to /v1/memory/insert with correct body and headers', async () => {
      fetchMock.mockResolvedValueOnce({
        ok: true,
        text: () =>
          Promise.resolve(
            JSON.stringify({
              success: true,
              data: { status: 'ok', stats: {} },
            })
          ),
      });
      const client = new TinyHumansMemoryClient({ token, baseUrl });
      const params: InsertMemoryParams = {
        title: 'My Doc',
        content: 'Some content',
        namespace: 'default',
        documentId: 'doc_1',
        sourceType: 'doc',
        metadata: { source: 'test' },
      };
      await client.insertMemory(params);
      expect(fetchMock).toHaveBeenCalledTimes(1);
      expect(fetchMock).toHaveBeenCalledWith(
        `${baseUrl}/v1/memory/insert`,
        expect.objectContaining({
          method: 'POST',
          headers: {
            'Content-Type': 'application/json',
            Authorization: `Bearer ${token}`,
          },
          body: JSON.stringify({
            title: 'My Doc',
            content: 'Some content',
            namespace: 'default',
            documentId: 'doc_1',
            sourceType: 'doc',
            metadata: { source: 'test' },
            priority: undefined,
            createdAt: undefined,
            updatedAt: undefined,
          }),
        })
      );
    });

    it('returns parsed response data', async () => {
      const data = { status: 'ok', stats: { ingested: 1 }, usage: { cost_usd: 0.01 } };
      fetchMock.mockResolvedValueOnce({
        ok: true,
        text: () => Promise.resolve(JSON.stringify({ success: true, data })),
      });
      const client = new TinyHumansMemoryClient({ token, baseUrl });
      const result = await client.insertMemory({
        title: 'T',
        content: 'C',
        namespace: 'ns',
        documentId: 'doc_2',
      });
      expect(result.success).toBe(true);
      expect(result.data).toEqual(data);
    });
  });

  describe('queryMemory', () => {
    it('validates query is required', async () => {
      const client = new TinyHumansMemoryClient({ token, baseUrl });
      await expect(
        client.queryMemory({ query: '' } as QueryMemoryParams)
      ).rejects.toThrow('query is required');
    });

    it('validates maxChunks range', async () => {
      const client = new TinyHumansMemoryClient({ token, baseUrl });
      fetchMock.mockResolvedValue({ ok: true, text: () => Promise.resolve('{"success":true,"data":{}}') });
      await expect(client.queryMemory({ query: 'q', maxChunks: 0 })).rejects.toThrow('maxChunks');
      await expect(client.queryMemory({ query: 'q', maxChunks: 201 })).rejects.toThrow('maxChunks');
    });

    it('POSTs to /v1/memory/query', async () => {
      fetchMock.mockResolvedValueOnce({
        ok: true,
        text: () =>
          Promise.resolve(
            JSON.stringify({
              success: true,
              data: { cached: false, context: {}, usage: {} },
            })
          ),
      });
      const client = new TinyHumansMemoryClient({ token, baseUrl });
      await client.queryMemory({ query: 'hello', namespace: 'ns', maxChunks: 10 });
      expect(fetchMock).toHaveBeenCalledWith(
        `${baseUrl}/v1/memory/query`,
        expect.objectContaining({
          method: 'POST',
          body: JSON.stringify({
            query: 'hello',
            includeReferences: undefined,
            namespace: 'ns',
            maxChunks: 10,
            documentIds: undefined,
            llmQuery: undefined,
          }),
        })
      );
    });
  });

  describe('deleteMemory', () => {
    it('POSTs to /v1/memory/admin/delete', async () => {
      fetchMock.mockResolvedValueOnce({
        ok: true,
        text: () =>
          Promise.resolve(
            JSON.stringify({
              success: true,
              data: {
                status: 'ok',
                userId: 'u1',
                nodesDeleted: 5,
                message: 'Deleted',
              },
            })
          ),
      });
      const client = new TinyHumansMemoryClient({ token, baseUrl });
      const params: DeleteMemoryParams = { namespace: 'my-ns' };
      await client.deleteMemory(params);
      expect(fetchMock).toHaveBeenCalledWith(
        `${baseUrl}/v1/memory/admin/delete`,
        expect.objectContaining({
          method: 'POST',
          body: JSON.stringify({ namespace: 'my-ns' }),
        })
      );
    });
  });

  describe('recallMemory', () => {
    it('validates maxChunks when provided', async () => {
      const client = new TinyHumansMemoryClient({ token, baseUrl });
      await expect(recallMemory({ maxChunks: 0 })).rejects.toThrow('maxChunks');
      await expect(recallMemory({ maxChunks: 1.5 })).rejects.toThrow('maxChunks');
      async function recallMemory(p: RecallMemoryParams) {
        return client.recallMemory(p);
      }
    });

    it('POSTs to /v1/memory/recall', async () => {
      fetchMock.mockResolvedValueOnce({
        ok: true,
        text: () =>
          Promise.resolve(
            JSON.stringify({
              success: true,
              data: { cached: false, response: 'context here' },
            })
          ),
      });
      const client = new TinyHumansMemoryClient({ token, baseUrl });
      await client.recallMemory({ namespace: 'ns', maxChunks: 10 });
      expect(fetchMock).toHaveBeenCalledWith(
        `${baseUrl}/v1/memory/recall`,
        expect.objectContaining({
          method: 'POST',
          body: JSON.stringify({ namespace: 'ns', maxChunks: 10 }),
        })
      );
    });
  });

  describe('recallMemories', () => {
    it('validates topK and minRetention when provided', async () => {
      const client = new TinyHumansMemoryClient({ token, baseUrl });
      await expect(recallMemories({ topK: 0 })).rejects.toThrow('topK');
      await expect(recallMemories({ minRetention: -1 })).rejects.toThrow('minRetention');
      async function recallMemories(p: RecallMemoriesParams) {
        return client.recallMemories(p);
      }
    });

    it('POSTs to /v1/memory/memories/recall', async () => {
      fetchMock.mockResolvedValueOnce({
        ok: true,
        text: () =>
          Promise.resolve(
            JSON.stringify({
              success: true,
              data: { memories: [] },
            })
          ),
      });
      const client = new TinyHumansMemoryClient({ token, baseUrl });
      await client.recallMemories({ namespace: 'ns', topK: 5 });
      expect(fetchMock).toHaveBeenCalledWith(
        `${baseUrl}/v1/memory/memories/recall`,
        expect.objectContaining({
          method: 'POST',
          body: JSON.stringify({ namespace: 'ns', topK: 5, minRetention: undefined, asOf: undefined }),
        })
      );
    });
  });

  describe('error handling', () => {
    it('throws AlphahumanError on non-ok response with error message', async () => {
      fetchMock.mockResolvedValueOnce({
        ok: false,
        status: 400,
        text: () => Promise.resolve(JSON.stringify({ success: false, error: 'Bad request' })),
      });
      const client = new TinyHumansMemoryClient({ token, baseUrl });
      try {
        await client.insertMemory({ title: 'T', content: 'C', namespace: 'ns', documentId: 'doc_3' });
        expect.fail('should throw');
      } catch (e) {
        expect(e).toBeInstanceOf(TinyHumansError);
        expect((e as TinyHumansError).message).toBe('Bad request');
        expect((e as TinyHumansError).status).toBe(400);
        expect((e as TinyHumansError).body).toEqual({ success: false, error: 'Bad request' });
      }
    });

    it('throws on non-JSON response', async () => {
      fetchMock.mockResolvedValueOnce({ ok: false, status: 500, text: () => Promise.resolve('not json') });
      const client = new TinyHumansMemoryClient({ token, baseUrl });
      await expect(client.insertMemory({ title: 'T', content: 'C', namespace: 'ns', documentId: 'doc_4' })).rejects.toThrow(
        'non-JSON response'
      );
    });
  });

  describe('chatMemory', () => {
    it('POSTs to /v1/memory/chat', async () => {
      fetchMock.mockResolvedValueOnce({
        ok: true,
        text: () => Promise.resolve(JSON.stringify({ success: true, data: { content: 'hello' } })),
      });
      const client = new TinyHumansMemoryClient({ token, baseUrl });
      await client.chatMemory({ messages: [{ role: 'user', content: 'hi' }] });
      expect(fetchMock).toHaveBeenCalledWith(
        `${baseUrl}/v1/memory/chat`,
        expect.objectContaining({
          method: 'POST',
          body: JSON.stringify({ messages: [{ role: 'user', content: 'hi' }] }),
        })
      );
    });
  });

  describe('listDocuments', () => {
    it('GETs from /v1/memory/documents with query params', async () => {
      fetchMock.mockResolvedValueOnce({
        ok: true,
        text: () => Promise.resolve(JSON.stringify({ success: true, data: {} })),
      });
      const client = new TinyHumansMemoryClient({ token, baseUrl });
      await client.listDocuments({ namespace: 'ns', limit: 10 });
      expect(fetchMock).toHaveBeenCalledWith(
        `${baseUrl}/v1/memory/documents?namespace=ns&limit=10`,
        expect.objectContaining({ method: 'GET' })
      );
    });
  });
});

File: packages/sdk-rust/tests/client_test.rs
//! Integration tests for TinyHumansMemoryClient using mockito.

use mockito::Server;
use tinyhumansai::{
    ChatMessage, DeleteMemoryParams, InsertMemoryParams, MemoryChatParams, QueryMemoryParams,
    RecallMemoriesParams, RecallMemoryParams, TinyHumanConfig, TinyHumansError,
    TinyHumansMemoryClient,
};

#[tokio::test]
async fn client_requires_token() {
    assert!(TinyHumansMemoryClient::new(TinyHumanConfig::new("")).is_err());
    assert!(TinyHumansMemoryClient::new(TinyHumanConfig::new("   ")).is_err());
}

#[tokio::test]
async fn insert_memory_validates_required() {
    let server = Server::new_async().await;
    let url = server.url();
    let client =
        TinyHumansMemoryClient::new(TinyHumanConfig::new("token").with_base_url(url)).unwrap();

    let empty_title = InsertMemoryParams {
        title: String::new(),
        content: "c".into(),
        namespace: "ns".into(),
        ..Default::default()
    };
    assert!(client.insert_memory(empty_title).await.is_err());

    let empty_ns = InsertMemoryParams {
        title: "t".into(),
        content: "c".into(),
        namespace: String::new(),
        ..Default::default()
    };
    assert!(client.insert_memory(empty_ns).await.is_err());

    let empty_doc_id = InsertMemoryParams {
        title: "t".into(),
        content: "c".into(),
        namespace: "ns".into(),
        document_id: String::new(),
        ..Default::default()
    };
    assert!(client.insert_memory(empty_doc_id).await.is_err());
}

#[tokio::test]
async fn insert_memory_posts_correctly() {
    let mut server = Server::new_async().await;
    let mock = server
        .mock("POST", "/memory/insert")
        .with_status(200)
        .with_header("content-type", "application/json")
        .with_body(r#"{"success":true,"data":{"status":"ok","stats":{}}}"#)
        .create_async()
        .await;

    let client =
        TinyHumansMemoryClient::new(TinyHumanConfig::new("test-token").with_base_url(server.url()))
            .unwrap();

    let params = InsertMemoryParams {
        title: "Doc".into(),
        content: "Content".into(),
        namespace: "default".into(),
        document_id: "doc-1".into(),
        ..Default::default()
    };
    let res = client.insert_memory(params).await.unwrap();
    assert!(res.success);
    assert_eq!(res.data.status.as_deref(), Some("ok"));
    mock.assert_async().await;
}

#[tokio::test]
async fn query_memory_validates_query() {
    let server = Server::new_async().await;
    let client =
        TinyHumansMemoryClient::new(TinyHumanConfig::new("token").with_base_url(server.url()))
            .unwrap();
    let params = QueryMemoryParams {
        query: String::new(),
        ..Default::default()
    };
    assert!(client.query_memory(params).await.is_err());
}

#[tokio::test]
async fn query_memory_validates_max_chunks() {
    let server = Server::new_async().await;
    let client =
        TinyHumansMemoryClient::new(TinyHumanConfig::new("token").with_base_url(server.url()))
            .unwrap();
    let p0 = QueryMemoryParams {
        query: "q".into(),
        max_chunks: Some(0.0),
        ..Default::default()
    };
    assert!(client.query_memory(p0).await.is_err());
    let pneg = QueryMemoryParams {
        query: "q".into(),
        max_chunks: Some(-1.0),
        ..Default::default()
    };
    assert!(client.query_memory(pneg).await.is_err());
}

#[tokio::test]
async fn delete_memory_posts_correctly() {
    let mut server = Server::new_async().await;
    let mock = server
        .mock("POST", "/memory/admin/delete")
        .with_status(200)
        .with_header("content-type", "application/json")
        .with_body(r#"{"success":true,"data":{"status":"ok","userId":"u1","nodesDeleted":5,"message":"Done"}}"#)
        .create_async()
        .await;

    let client =
        TinyHumansMemoryClient::new(TinyHumanConfig::new("token").with_base_url(server.url()))
            .unwrap();
    let res = client
        .delete_memory(DeleteMemoryParams {
            namespace: Some("ns".into()),
        })
        .await
        .unwrap();
    assert!(res.success);
    assert_eq!(res.data.nodes_deleted, 5);
    mock.assert_async().await;
}

#[tokio::test]
async fn recall_memory_posts_correctly() {
    let mut server = Server::new_async().await;
    let mock = server
        .mock("POST", "/memory/recall")
        .with_status(200)
        .with_header("content-type", "application/json")
        .with_body(r#"{"success":true,"data":{"cached":false,"response":"context"}}"#)
        .create_async()
        .await;

    let client =
        TinyHumansMemoryClient::new(TinyHumanConfig::new("token").with_base_url(server.url()))
            .unwrap();
    let res = client
        .recall_memory(RecallMemoryParams {
            namespace: Some("ns".into()),
            max_chunks: Some(10.0),
        })
        .await
        .unwrap();
    assert!(res.success);
    assert_eq!(res.data.response.as_deref(), Some("context"));
    mock.assert_async().await;
}

#[tokio::test]
async fn recall_memories_posts_correctly() {
    let mut server = Server::new_async().await;
    let mock = server
        .mock("POST", "/memory/memories/recall")
        .with_status(200)
        .with_header("content-type", "application/json")
        .with_body(r#"{"success":true,"data":{"memories":[]}}"#)
        .create_async()
        .await;

    let client =
        TinyHumansMemoryClient::new(TinyHumanConfig::new("token").with_base_url(server.url()))
            .unwrap();
    let res = client
        .recall_memories(RecallMemoriesParams {
            namespace: Some("ns".into()),
            top_k: Some(5.0),
            ..Default::default()
        })
        .await
        .unwrap();
    assert!(res.success);
    assert!(res.data.memories.is_empty());
    mock.assert_async().await;
}

#[tokio::test]
async fn api_error_returns_tinyhumans_error() {
    let mut server = Server::new_async().await;
    server
        .mock("POST", "/memory/insert")
        .with_status(400)
        .with_header("content-type", "application/json")
        .with_body(r#"{"success":false,"error":"Bad request"}"#)
        .create_async()
        .await;

    let client =
        TinyHumansMemoryClient::new(TinyHumanConfig::new("token").with_base_url(server.url()))
            .unwrap();
    let params = InsertMemoryParams {
        title: "T".into(),
        content: "C".into(),
        namespace: "ns".into(),
        document_id: "doc-1".into(),
        ..Default::default()
    };
    let err = client.insert_memory(params).await.unwrap_err();
    match &err {
        TinyHumansError::Api {
            message, status, ..
        } => {
            assert_eq!(*status, 400);
            assert!(message.contains("Bad request"));
        }
        _ => panic!("expected Api error, got {:?}", err),
    }
}

#[tokio::test]
async fn memory_chat_posts_correctly() {
    let mut server = Server::new_async().await;
    let mock = server
        .mock("POST", "/memory/chat")
        .with_status(200)
        .with_header("content-type", "application/json")
        .with_body(
            r#"{"success":true,"data":{"content":"hello","usage":{"prompt_tokens":3,"completion_tokens":1,"total_tokens":4},"model":"gpt-test"}}"#,
        )
        .create_async()
        .await;

    let client =
        TinyHumansMemoryClient::new(TinyHumanConfig::new("token").with_base_url(server.url()))
            .unwrap();
    let res = client
        .memory_chat(MemoryChatParams {
            messages: vec![ChatMessage {
                role: "user".into(),
                content: "hello".into(),
            }],
            ..Default::default()
        })
        .await
        .unwrap();

    assert!(res.success);
    assert_eq!(res.data.content.as_deref(), Some("hello"));
    mock.assert_async().await;
}
